samples/runtime/qpic-simulator/simulator/Processor.cs

```diff
@@ -0,0 +1,186 @@
+            // get all Qubit objects from args
+            var qubitIds = args.Select(obj => obj switch { Qubit q => q, _ => null })
+                               .WhereNotNull();
+
+            foreach (var builder in Pictures.Values) {
+                builder.AppendFormat(QpicFormatter.Instance, format, args);
+                foreach (var (qubit, invert) in classicallyControlledQubits.Values) {
+                    if (!qubitIds.Contains(qubit)) {
+                        builder.AppendFormat(QpicFormatter.Instance, " {1}{0}", qubit, invert ? "-" : "");
+                    }
+                }
+                builder.AppendLine();
+            }
+        }
+
+        #region Quantum operations
+        public override void OnAllocateQubits(IQArray<Qubit> qubits) {
+            foreach (var qubit in qubits) {
+                AddCommand("{0} W", qubit);
+            }
+        }
+
+        public override void X(Qubit qubit) {
+            AddCommand("{0} X", qubit);
+        }
+
+        public override void ControlledX(IQArray<Qubit> controls, Qubit qubit) {
+            var format = String.Join(" ", Enumerable.Range(0, (int)controls.Length).Select(i => $"{{{i}}}")) + $" +{{{controls.Length}}}";
+
+            AddCommand(format, controls.Append(qubit).ToArray<object>());
+        }
+
+        public override void Y(Qubit qubit) {
+            AddCommand("{0} Y", qubit);
+        }
+
+        public override void ControlledY(IQArray<Qubit> controls, Qubit qubit) {
+            var format = "{{{0}}} Y " + String.Join(" ", Enumerable.Range(0, (int)controls.Length).Select(i => $"{{{i + 1}}}"));
+
+            AddCommand(format, controls.Append(qubit).ToArray<object>());
+        }
+
+        public override void Z(Qubit qubit) {
+            AddCommand("{0} Z", qubit);
+        }
+
+        public override void ControlledZ(IQArray<Qubit> controls, Qubit qubit) {
+            var format = "{{{0}}} Z " + String.Join(" ", Enumerable.Range(0, (int)controls.Length).Select(i => $"{{{i + 1}}}"));
+
+            AddCommand(format, controls.Prepend(qubit).ToArray<object>());
+        }
+
+        public override void H(Qubit qubit) {
+            AddCommand("{0} H", qubit);
+        }
+
+        public override void S(Qubit qubit) {
+            AddCommand("{0} G {{$S$}}", qubit);
+        }
+
+        public override void SAdjoint(Qubit qubit) {
+            AddCommand("{0} G {{$S^\\dagger$}}", qubit);
+        }
+
+        public override void T(Qubit qubit) {
+            AddCommand("{0} G {{$T$}}", qubit);
+        }
+
+        public override void TAdjoint(Qubit qubit) {
+            AddCommand("{0} G {{$T^\\dagger$}}", qubit);
+        }
+
+        public override void ExpFrac(IQArray<Pauli> paulis, long numerator, long power, IQArray<Qubit> qubits) {
+            // we are using a color coding for some angles according to the visualizations in
+            // [arXiv:1808.02892](https://arxiv.org/abs/1808.02892)
```

Comment:
I think that this is a good idea, however, in this case it is intentional to use the exact same colours which are used in the paper, since there's no other indication on which angle is being used. Another solution is to use a different way to visualise `ExpFrac` gates in `qpic`.

Reply:
Maybe a toggle or something to switch between color schemes? Using color blind safe colors is an extremely helpful accessibility feature I have found, and pretty easy to address as @cgranade points out.

Reply:
Is there a way to add this information as text? Carrying info in colors only seems so fragile, regardless of the encoding used - once my laptop switches to night mode the circuit changes.

Reply:
@cgranade @crazy4pi314 @tcNickolas Thanks to everyone for your feedback. For now I decided to drop the visualisation for ExpFrac, because there to the best of my knowledge there is currently no good way for `qpic` to support these gates, let alone some common approach to visualise exponentiated Pauli gates.